File: src/main/scala/nu/tengstrand/tetrisanalyzer/piecemove/ValidPieceMovesForEmptyBoard.scala
package nu.tengstrand.tetrisanalyzer.piecemove

import nu.tengstrand.tetrisanalyzer.settings.GameSettings
import nu.tengstrand.tetrisanalyzer.piece.Piece
import nu.tengstrand.tetrisanalyzer.board.Board

/**
 * Calculates all valid moves for a given piece on an empty board.
 * If sliding is on, rotations and lateral movements can only occur in the top row.
 */
class ValidPieceMovesForEmptyBoard(val board: Board, val piece: Piece, settings: GameSettings, isSlidingEnabled: Boolean = false) {
  private val boardWidth = board.width
  private val boardHeight = board.height
  private var visitedPieceMoves = new VisitedPieceMoves(board, piece)
  private val rotationDirection = settings.rotationDirection

  private def markAsVisited(fromMovement: Movement, movement: Movement) {
    visitedPieceMoves.visit(movement)
    movement.linkTo(fromMovement)
  }

  private def isPieceInsideBoard(movement: Movement): Boolean = {
    val move = movement.pieceMove.move
    (move.x >= 0 && move.x + piece.width(move.rotation) <= boardWidth &&
      move.y >= 0 && move.y + piece.height(move.rotation) <= boardHeight)
  }

  /**
   * Calculates all valid moves as a linked list of moves by returning
   * the starting piece move for the board.
   */
  def startMove: PieceMove = {
    val startMove = settings.pieceStartMove(boardWidth, piece)
    val startMovement = new Movement(PieceMove(board, piece, startMove))
    val fromMovement = new Movement(PieceMove(board, piece, startMove.up))

    calculateValidMoves(fromMovement, startMovement)

    if (fromMovement.pieceMove.down == null)
      throw new IllegalStateException("Illegal start position for piece")

    fromMovement.pieceMove.down
  }

  private def calculateValidMoves(fromMovement: Movement, movement: Movement) {
    while (visitedPieceMoves.isUnvisited(movement) && isPieceInsideBoard(movement)) {
      markAsVisited(fromMovement, movement)
      if (isSlidingEnabled || movement.pieceMove.move.y == 0) {
        val moveAdjustment = settings.rotationAdjustment(movement.pieceMove.piece, movement.pieceMove.move.rotation)
        calculateValidMoves(movement, movement.rotate(moveAdjustment, rotationDirection, piece.rotationModulus, visitedPieceMoves))
        calculateValidMoves(movement, movement.left(visitedPieceMoves))
        calculateValidMoves(movement, movement.right(visitedPieceMoves))
      }
      calculateValidMoves(movement, movement.down(visitedPieceMoves))
    }
  }
}

File: langs/java/src/main/java/com/github/tetrisanalyzer/piecemove/ValidPieceMovesForEmptyBoard.java
package com.github.tetrisanalyzer.piecemove;

import com.github.tetrisanalyzer.board.Board;
import com.github.tetrisanalyzer.move.Move;
import com.github.tetrisanalyzer.move.rotation.RotationDirection;
import com.github.tetrisanalyzer.piece.Piece;
import com.github.tetrisanalyzer.settings.GameSettings;

/**
 * Calculates all valid moves for a given piece on an empty board.
 * If sliding is on, rotations and lateral movements can only occur in the top row.
 */
public class ValidPieceMovesForEmptyBoard {
    private boolean isSlidingOn;
    private Board board;
    private Piece piece;
    private GameSettings settings;
    private VisitedPieceMoves visitedPieceMoves;
    private RotationDirection rotationDirection;

    public ValidPieceMovesForEmptyBoard(Board board, Piece piece, GameSettings settings) {
        this.board = board;
        this.piece = piece;
        this.settings = settings;
        visitedPieceMoves = new VisitedPieceMoves(board, piece);
        rotationDirection = settings.rotationDirection();
        isSlidingOn = settings.isSlidingEnabled();
    }

    private void markAsVisited(Movement fromMovement, Movement movement) {
        visitedPieceMoves.visit(movement);
        movement.linkTo(fromMovement);
    }

    /**
     * Calculates all valid moves as a linked list of moves by returning the starting piece move for the board.
     */
    public PieceMove calculateStartMove() {
        Move startMove = new Move(0, settings.pieceStartX() + piece.startX(), settings.pieceStartY + piece.startY());
        Movement startMovement = new Movement(new PieceMove(board, piece, startMove));
        Movement fromMovement = new Movement(new PieceMove(board, piece, startMove.up()));

        if (!isSlidingOn) {
            // When sliding is off, ensure that the piece can perform all valid rotations in its starting position.
            while (startMovement.pieceMove.isFree() && !isAllRotationsFree(startMovement)) {
                fromMovement = startMovement;
                startMovement = startMovement.down(visitedPieceMoves);
                markAsVisited(fromMovement, startMovement);
            }
        }

        ensureStartingPositionIsInsideBoard(startMovement);
        calculateValidMoves(fromMovement, startMovement, true);

        return startMovement.pieceMove;
    }

    /**
     * Calculate all valid moves on an empty board recursively.
     *
     * @param fromMovement the position and orientation we are coming from
     * @param movement position and orientation of current piece
     * @param firstRow true if the action "drop piece" has started
     */
    private void calculateValidMoves(Movement fromMovement, Movement movement, boolean firstRow) {
        while (visitedPieceMoves.isUnvisited(movement) && movement.isPieceInsideBoard()) {
            markAsVisited(fromMovement, movement);
            if (isSlidingOn || firstRow) {
                calculateValidMoves(movement, movement.rotate(rotationDirection, visitedPieceMoves), firstRow);
                calculateValidMoves(movement, movement.left(visitedPieceMoves), firstRow);
                calculateValidMoves(movement, movement.right(visitedPieceMoves), firstRow);
            }
            calculateValidMoves(movement, movement.down(visitedPieceMoves), false);
        }
    }

    private boolean isAllRotationsFree(Movement movement) {
        for (int i=0; i<=piece.rotationsEndIndex(); i++) {
            if (!movement.isPieceInsideBoard() || !movement.pieceMove.isFree()) {
                return false;
            }
            movement = movement.rotate(rotationDirection, visitedPieceMoves);
        }
        return true;
    }

    private void ensureStartingPositionIsInsideBoard(Movement startMovement) {
        if (!startMovement.isPieceInsideBoard()) {
            throw new IllegalStateException("The start piece position is outside the board, " +
                    "start piece settings: [" + settings.pieceStartX + "," + settings.pieceStartY + "]" +
                    ", piece: " + piece +
                    ", piece adjustment (rotation,x, y): " + startMovement.pieceMove.move);
        }
    }
}
